{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set-up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/users/ymo/.local/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import yaml\n",
    "import logging\n",
    "import mudata\n",
    "import pandas as pd\n",
    "\n",
    "# Change path to wherever you have repo locally\n",
    "sys.path.append('/oak/stanford/groups/engreitz/Users/ymo/Tools/cNMF_benchmarking/cNMF_benchmarking_pipeline')\n",
    "\n",
    "from Evaluation.src import (\n",
    "    compute_categorical_association,\n",
    "    compute_geneset_enrichment,\n",
    "    compute_trait_enrichment,\n",
    "    compute_perturbation_association,\n",
    "    #compute_explained_variance_ratio,\n",
    "    compute_motif_enrichment\n",
    ")\n",
    "from Evaluation.src.enrichment_trait import process_enrichment_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# categorical association, geneset enrichment, GO term enrichment, trait enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Performing tests at single-cell level. Significance will likely be inflated\n",
      "Testing sample association: 100%|██████████| 30/30 [00:03<00:00,  9.59programs/s]\n",
      "Identifying differential sample: 100%|██████████| 30/30 [00:03<00:00,  7.85programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0002 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 30/30 [00:09<00:00,  3.14programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0001 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 30/30 [00:25<00:00,  1.17programs/s]\n",
      "Running Fisher enrichment: 100%|██████████| 30/30 [00:02<00:00, 13.33programs/s]\n",
      "INFO:root:Performing tests at single-cell level. Significance will likely be inflated\n",
      "Testing sample association: 100%|██████████| 60/60 [00:06<00:00,  9.43programs/s]\n",
      "Identifying differential sample: 100%|██████████| 60/60 [00:07<00:00,  7.90programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0002 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 60/60 [00:16<00:00,  3.54programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0001 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 60/60 [00:48<00:00,  1.23programs/s]\n",
      "Running Fisher enrichment: 100%|██████████| 60/60 [00:04<00:00, 13.60programs/s]\n",
      "INFO:root:Performing tests at single-cell level. Significance will likely be inflated\n",
      "Testing sample association: 100%|██████████| 80/80 [00:08<00:00,  9.42programs/s]\n",
      "Identifying differential sample: 100%|██████████| 80/80 [00:10<00:00,  7.86programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0002 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 80/80 [00:22<00:00,  3.55programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0001 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 80/80 [01:05<00:00,  1.23programs/s]\n",
      "Running Fisher enrichment: 100%|██████████| 80/80 [00:05<00:00, 13.76programs/s]\n",
      "INFO:root:Performing tests at single-cell level. Significance will likely be inflated\n",
      "Testing sample association: 100%|██████████| 100/100 [00:10<00:00,  9.25programs/s]\n",
      "Identifying differential sample: 100%|██████████| 100/100 [00:12<00:00,  7.70programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0002 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 100/100 [00:28<00:00,  3.51programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0001 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 100/100 [01:21<00:00,  1.22programs/s]\n",
      "Running Fisher enrichment: 100%|██████████| 100/100 [00:07<00:00, 13.85programs/s]\n",
      "INFO:root:Performing tests at single-cell level. Significance will likely be inflated\n",
      "Testing sample association: 100%|██████████| 200/200 [00:23<00:00,  8.46programs/s]\n",
      "Identifying differential sample: 100%|██████████| 200/200 [00:27<00:00,  7.20programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0002 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 200/200 [00:57<00:00,  3.50programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0001 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 200/200 [02:44<00:00,  1.21programs/s]\n",
      "Running Fisher enrichment: 100%|██████████| 200/200 [00:14<00:00, 14.21programs/s]\n",
      "INFO:root:Performing tests at single-cell level. Significance will likely be inflated\n",
      "Testing sample association: 100%|██████████| 250/250 [00:35<00:00,  6.99programs/s]\n",
      "Identifying differential sample: 100%|██████████| 250/250 [00:38<00:00,  6.54programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0002 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 250/250 [01:12<00:00,  3.45programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0001 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 250/250 [03:27<00:00,  1.20programs/s]\n",
      "Running Fisher enrichment: 100%|██████████| 250/250 [00:17<00:00, 14.13programs/s]\n",
      "INFO:root:Performing tests at single-cell level. Significance will likely be inflated\n",
      "Testing sample association: 100%|██████████| 300/300 [00:35<00:00,  8.54programs/s]\n",
      "Identifying differential sample: 100%|██████████| 300/300 [00:40<00:00,  7.43programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0002 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 300/300 [01:29<00:00,  3.36programs/s]\n",
      "INFO:root:Downloading and generating Enrichr library gene sets...\n",
      "INFO:root:0001 gene_sets have been filtered out when max_size=2000 and min_size=0\n",
      "Running Fisher enrichment: 100%|██████████| 300/300 [04:22<00:00,  1.14programs/s]\n",
      "Running Fisher enrichment: 100%|██████████| 300/300 [00:21<00:00, 14.20programs/s]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "folder = \"/oak/stanford/groups/engreitz/Users/ymo/NMF_re-inplementing/Results/torch-cNMF_evaluation/100k_cells_10iter_torch_mu_batch\"\n",
    "mdata_guide = \"/oak/stanford/groups/engreitz/Users/ymo/Tools/cNMF_benchmarking/cNMF_benchmarking_pipeline/Evaluation/Resources/mdata_guide.h5mu\"\n",
    "\n",
    "def assign_guide(mdatam,file):\n",
    "\n",
    "    # read mdata with guide\n",
    "    mdata_guide = mudata.read(file)\n",
    "\n",
    "    mdata['cNMF'].uns[\"guide_names\"] = mdata_guide[\"cNMF_100\"].uns[\"guide_names\"]\n",
    "    mdata['cNMF'].uns[\"guide_targets\"] = mdata_guide[\"cNMF_100\"].uns[\"guide_targets\"]\n",
    "    mdata['cNMF'].obsm[\"guide_assignment\"] = mdata_guide[\"cNMF_100\"].obsm[\"guide_assignment\"]\n",
    "\n",
    "\n",
    "for k in [30, 60, 80, 100, 200, 250, 300]:  \n",
    "\n",
    "    os.makedirs(f\"{folder}/Eval/{k}\", exist_ok=True)\n",
    "\n",
    "    output_folder = f\"{folder}/Eval/{k}\"\n",
    "\n",
    "    # Load mdata\n",
    "    mdata = mudata.read('{}/adata/cNMF_{}_2_0.h5mu'.format(folder,k))\n",
    "    assign_guide(mdata, mdata_guide) \n",
    "\n",
    "\n",
    "    # Run categorical assocation\n",
    "    results_df, posthoc_df = compute_categorical_association(mdata, prog_key='cNMF', categorical_key='sample', \n",
    "                                                            pseudobulk_key=None, test='dunn', n_jobs=-1, inplace=False)\n",
    "\n",
    "    results_df.to_csv('{}/{}_categorical_association_results.txt'.format(output_folder,k), sep='\\t', index=False) # This was made wide form to insert into .var of the program anndata.\n",
    "    posthoc_df.to_csv('{}/{}_categorical_association_posthoc.txt'.format(output_folder,k), sep='\\t', index=False)\n",
    "\n",
    " \n",
    "    # Run perturbation assocation\n",
    "    for samp in mdata['rna'].obs['sample'].unique():\n",
    "        mdata_ = mdata[mdata['rna'].obs['sample']==samp]\n",
    "        test_stats_df = compute_perturbation_association(mdata_, prog_key='cNMF', \n",
    "                                                        collapse_targets=True,\n",
    "                                                        pseudobulk=False,\n",
    "                                                        reference_targets=('non-targeting'),\n",
    "                                                        n_jobs=-1, inplace=False)\n",
    "\n",
    "        test_stats_df.to_csv('{}/{}_perturbation_association_results_{}.txt'.format(folder,k,samp), sep='\\t', index=False)\n",
    "  \n",
    "\n",
    "    # Gene-set enrichment\n",
    "    pre_res = compute_geneset_enrichment(mdata, prog_key='cNMF', data_key='rna', prog_nam=None,\n",
    "                                        organism='human', library='Reactome_2022', method=\"fisher\",\n",
    "                                        database='enrichr', loading_rank_thresh=300, n_jobs=-1, \n",
    "                                        inplace=False, user_geneset=None)\n",
    "    pre_res.to_csv('{}/{}_geneset_enrichment.txt'.format(output_folder,k), sep='\\t', index=False)\n",
    "\n",
    "    # GO Term enrichment\n",
    "    pre_res = compute_geneset_enrichment(mdata, prog_key='cNMF', data_key='rna', prog_nam=None,\n",
    "                                        organism='human', library='GO_Biological_Process_2023', method=\"fisher\",\n",
    "                                        database='enrichr', loading_rank_thresh=300, n_jobs=-1, \n",
    "                                        inplace=False, user_geneset=None)\n",
    "    pre_res.to_csv('{}/{}_GO_term_enrichment.txt'.format(output_folder,k), sep='\\t', index=False)\n",
    "\n",
    "    # Run trait enrichment\n",
    "    pre_res_trait = compute_trait_enrichment(mdata, gwas_data='/oak/stanford/groups/engreitz/Users/ymo/Tools/cNMF_benchmarking/cNMF_benchmarking_pipeline/Evaluation/Resources/OpenTargets_L2G_Filtered.csv.gz', \n",
    "                                            prog_key='cNMF', prog_nam=None, data_key='rna', \n",
    "                                            library='OT_GWAS', n_jobs=-1, inplace=False, \n",
    "                                            key_column='trait_efos', gene_column='gene_name', \n",
    "                                            method='fisher', loading_rank_thresh=300)\n",
    "    pre_res_trait.to_csv('{}/{}_trait_enrichment.txt'.format(output_folder,k), sep='\\t', index=False)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Motif Enrichment (compliation working on progress)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Format files\n",
    "\n",
    "# Thresholds\n",
    "# score_thresh_abc_e2g_enhancer = 0.015\n",
    "# score_thresh_abc_e2g_promoter = 0.8\n",
    "\n",
    "# for i in range(4):\n",
    "#     e2g = pd.read_csv('scE2G_links/EnhancerPredictionsAllPutative.ForVariantOverlap.shrunk150bp_D{}.tsv'.format(i), sep='\\t')\n",
    "\n",
    "#     e2g_enhancers = e2g.loc[(e2g['class']!='promoter') &\\\n",
    "#                             (e2g['ABC.Score']>score_thresh_abc_e2g_enhancer)]\n",
    "#     e2g_enhancers = e2g_enhancers.loc[:,['chr', 'start', 'end', 'name', 'class', 'ABC.Score', 'TargetGene']]\n",
    "#     e2g_enhancers.columns = [        'chromosome', 'start', 'end', 'seq_name', 'seq_class', 'seq_score', 'gene_name']\n",
    "\n",
    "#     e2g_enhancers.to_csv('scE2G_links/EnhancerPredictionsAllPutative.ForVariantOverlap.shrunk150bp_D{}_enhancer.tsv'.format(i), \n",
    "#                           sep='\\t', index=False)\n",
    "\n",
    "#     e2g_promoters = e2g.loc[(e2g['class']=='promoter') &\\\n",
    "#                             (e2g['ABC.Score']>score_thresh_abc_e2g_promoter)]\n",
    "#     e2g_promoters = e2g_promoters.loc[:,['chr', 'start', 'end', 'name', 'class', 'ABC.Score', 'TargetGene']]\n",
    "#     e2g_promoters.columns = ['chromosome', 'start', 'end', 'seq_name', 'seq_class', 'seq_score', 'gene_name']\n",
    "\n",
    "#     e2g_promoters.to_csv('scE2G_links/EnhancerPredictionsAllPutative.ForVariantOverlap.shrunk150bp_D{}_promoter.tsv'.format(i), \n",
    "#                         sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Run in script\n",
    "\n",
    "\n",
    "fimo_thresh_enhancer = 1e-6\n",
    "fimo_thresh_promoter = 1e-4\n",
    "\n",
    "# Load program data\n",
    "mdata = mudata.read('/oak/stanford/groups/engreitz/Users/ymo/NMF_re-inplementing/Results/torch-cNMF_evaluation/090425_100k_cells_10iter_torch_mu_online/adata/cNMF_30_2.0.h5mu')\n",
    "output_path = '/oak/stanford/groups/engreitz/Users/ymo/NMF_re-inplementing/Results/torch-cNMF_evaluation/090425_100k_cells_10iter_torch_mu_online/Eval/30'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 92284 × 5451\n",
       "    obs: 'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'leiden'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    layers: 'norm10k'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata['rna']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:biothings.client:Input sequence provided is already in string format. No operation performed\n",
      "WARNING:biothings.client:Input sequence provided is already in string format. No operation performed\n",
      "INFO:biothings.client:querying 1-1000 ...\n",
      "INFO:httpx:HTTP Request: POST https://mygene.info/v3/query/ \"HTTP/1.1 200 OK\"\n",
      "INFO:biothings.client:querying 1001-2000 ...\n",
      "INFO:httpx:HTTP Request: POST https://mygene.info/v3/query/ \"HTTP/1.1 200 OK\"\n",
      "INFO:biothings.client:querying 2001-3000 ...\n",
      "INFO:httpx:HTTP Request: POST https://mygene.info/v3/query/ \"HTTP/1.1 200 OK\"\n",
      "INFO:biothings.client:querying 3001-4000 ...\n",
      "INFO:httpx:HTTP Request: POST https://mygene.info/v3/query/ \"HTTP/1.1 200 OK\"\n",
      "INFO:biothings.client:querying 4001-5000 ...\n",
      "INFO:httpx:HTTP Request: POST https://mygene.info/v3/query/ \"HTTP/1.1 200 OK\"\n",
      "INFO:biothings.client:querying 5001-5451 ...\n",
      "INFO:httpx:HTTP Request: POST https://mygene.info/v3/query/ \"HTTP/1.1 200 OK\"\n",
      "INFO:biothings.client:Finished.\n",
      "WARNING:biothings.client:5 input query terms found dup hits:\t[('ENSG00000227110', 2), ('ENSG00000228566', 2), ('ENSG00000234352', 2), ('ENSG00000249738', 2), ('E\n",
      "WARNING:biothings.client:110 input query terms found no hit:\t['ENSG00000130489', 'ENSG00000161149', 'ENSG00000167945', 'ENSG00000182912', 'ENSG00000186354', 'ENS\n",
      "INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n"
     ]
    }
   ],
   "source": [
    "import mygene\n",
    "\n",
    "def rename_gene(mdata):\n",
    "\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    gene_list = mdata['cNMF'].uns['var_names'].tolist()\n",
    "    annotations = mg.querymany(gene_list, scopes='ensembl.gene', fields='symbol', species='human')\n",
    "\n",
    "    # Process the results to create mapping\n",
    "    gene_dict = {}\n",
    "    \n",
    "    for item in annotations:\n",
    "        if 'symbol' in item:\n",
    "            gene_dict[item['query']] = item['symbol']\n",
    "\n",
    "    mdata['cNMF'].uns['var_names'] = [gene_dict.get(x, x) for x in mdata['cNMF'].uns['var_names']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "mdata.var_names_make_unique()  # if you need unique names first\n",
    "mdata.var.index = mdata.var['gene_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SEMA3F',\n",
       " 'CFTR',\n",
       " 'CYP51A1',\n",
       " 'HECW1',\n",
       " 'KLHL13',\n",
       " 'CASP10',\n",
       " 'CFLAR',\n",
       " 'TFPI',\n",
       " 'MTMR7',\n",
       " 'SLC7A2',\n",
       " 'ARF5',\n",
       " 'PLXND1',\n",
       " 'CAMKK1',\n",
       " 'ST7',\n",
       " 'CALCR',\n",
       " 'SLC25A5',\n",
       " 'THSD7A',\n",
       " 'ACSM3',\n",
       " 'SPPL2B',\n",
       " 'TSPOAP1',\n",
       " 'CROT',\n",
       " 'ITGA3',\n",
       " 'CRLF1',\n",
       " 'OSBPL7',\n",
       " 'TAC1',\n",
       " 'CACNA1G',\n",
       " 'TNFRSF12A',\n",
       " 'DLX6',\n",
       " 'MAP3K9',\n",
       " 'RALA',\n",
       " 'ETV1',\n",
       " 'USH1C',\n",
       " 'GTF2IRD1',\n",
       " 'ARHGAP44',\n",
       " 'UPP2',\n",
       " 'PROM1',\n",
       " 'CEACAM21',\n",
       " 'NOS2',\n",
       " 'GAS7',\n",
       " 'CACNA2D2',\n",
       " 'E2F2',\n",
       " 'CDKL5',\n",
       " 'NADK',\n",
       " 'ADAM22',\n",
       " 'CYB561',\n",
       " 'AASS',\n",
       " 'CRY1',\n",
       " 'ST3GAL1',\n",
       " 'REV3L',\n",
       " 'TENM1',\n",
       " 'PAX7',\n",
       " 'ETV7',\n",
       " 'CD9',\n",
       " 'GIPR',\n",
       " 'STAB1',\n",
       " 'SLC6A13',\n",
       " 'IDS',\n",
       " 'ZNF200',\n",
       " 'LRRC23',\n",
       " 'BTK',\n",
       " 'SCMH1',\n",
       " 'FYN',\n",
       " 'HIVEP2',\n",
       " 'LYPLA2',\n",
       " 'SLC6A7',\n",
       " 'TSPAN9',\n",
       " 'ABHD5',\n",
       " 'ANOS1',\n",
       " 'PLAUR',\n",
       " 'ANLN',\n",
       " 'MAP4K3',\n",
       " 'GABRA3',\n",
       " 'PRICKLE3',\n",
       " 'CLDN11',\n",
       " 'GPRC5A',\n",
       " 'MAMLD1',\n",
       " 'ACP3',\n",
       " 'MDH1',\n",
       " 'CCDC88C',\n",
       " 'ISL1',\n",
       " 'SLC38A5',\n",
       " 'RUFY3',\n",
       " 'CNTN1',\n",
       " 'WWTR1',\n",
       " 'ATP1A2',\n",
       " 'ZNF582',\n",
       " 'SNAI2',\n",
       " 'HGF',\n",
       " 'ADGRA2',\n",
       " 'SAMD4A',\n",
       " 'PLEKHB1',\n",
       " 'NRXN3',\n",
       " 'CPS1',\n",
       " 'FHL1',\n",
       " 'NLRP2',\n",
       " 'SLC45A4',\n",
       " 'GRAMD1B',\n",
       " 'DEPDC1',\n",
       " 'RRAGD',\n",
       " 'HSD17B6',\n",
       " 'VIM',\n",
       " 'RNASET2',\n",
       " 'CD44',\n",
       " 'PRKCH',\n",
       " 'POU2F2',\n",
       " 'BRD9',\n",
       " 'ANK1',\n",
       " 'HMGB3',\n",
       " 'IKZF2',\n",
       " 'ARHGAP31',\n",
       " 'LRRC7',\n",
       " 'FUT8',\n",
       " 'GAB2',\n",
       " 'APBA2',\n",
       " 'TMSB10',\n",
       " 'VCL',\n",
       " 'SLC18A1',\n",
       " 'FLT4',\n",
       " 'FBXO42',\n",
       " 'HOXC8',\n",
       " 'TLL1',\n",
       " 'VCAN',\n",
       " 'MSR1',\n",
       " 'CDH1',\n",
       " 'DNAH5',\n",
       " 'PHLPP2',\n",
       " 'SLC66A1',\n",
       " 'CTNS',\n",
       " 'CDH10',\n",
       " 'INPP4A',\n",
       " 'MYO16',\n",
       " 'PARP3',\n",
       " 'TNC',\n",
       " 'RIPOR3',\n",
       " 'JADE2',\n",
       " 'ZIC2',\n",
       " 'LCP2',\n",
       " 'HSPA5',\n",
       " 'GPM6B',\n",
       " 'PREX2',\n",
       " 'ARAP2',\n",
       " 'ANO2',\n",
       " 'ARHGAP6',\n",
       " 'HDAC9',\n",
       " 'RRM2B',\n",
       " 'LMO3',\n",
       " 'CELF2',\n",
       " 'COL9A2',\n",
       " 'KITLG',\n",
       " 'ADAMTS6',\n",
       " 'LTBP1',\n",
       " 'RCN1',\n",
       " 'ELN',\n",
       " 'RFC2',\n",
       " 'NEDD4L',\n",
       " 'NEXMIF',\n",
       " 'NFE2L3',\n",
       " 'SLC4A8',\n",
       " 'LAMC3',\n",
       " 'COL23A1',\n",
       " 'FHIP1B',\n",
       " 'HERPUD1',\n",
       " 'POLQ',\n",
       " 'FNIP2',\n",
       " 'MSMO1',\n",
       " 'FSTL4',\n",
       " 'FOXN3',\n",
       " 'METTL24',\n",
       " 'NNAT',\n",
       " 'MCF2L2',\n",
       " 'LAMA3',\n",
       " 'KCNQ1',\n",
       " 'OPN3',\n",
       " 'SDCCAG8',\n",
       " 'HHAT',\n",
       " 'FOXC1',\n",
       " 'TBC1D22A',\n",
       " 'ATP9A',\n",
       " 'FAM168A',\n",
       " 'RELT',\n",
       " 'MCOLN3',\n",
       " 'CCDC85A',\n",
       " 'NPFFR2',\n",
       " 'PHF21B',\n",
       " 'TRAF1',\n",
       " 'IL17RB',\n",
       " 'DCBLD2',\n",
       " 'PKP2',\n",
       " 'MSH4',\n",
       " 'PRDM1',\n",
       " 'TMCC3',\n",
       " 'LAMC2',\n",
       " 'CDK14',\n",
       " 'ATP2B4',\n",
       " 'DGKG',\n",
       " 'FLYWCH1',\n",
       " 'ALDH18A1',\n",
       " 'TARBP1',\n",
       " 'CDK17',\n",
       " 'DNAJC25',\n",
       " 'SLC2A3',\n",
       " 'STYK1',\n",
       " 'PTPRU',\n",
       " 'RIMBP2',\n",
       " 'COL11A1',\n",
       " 'HDAC7',\n",
       " 'LZTS1',\n",
       " 'PRDM6',\n",
       " 'WNT8A',\n",
       " 'TNK2',\n",
       " 'CDH3',\n",
       " 'ARSF',\n",
       " 'SEZ6',\n",
       " 'U2AF2',\n",
       " 'ADCK1',\n",
       " 'LIMCH1',\n",
       " 'ST3GAL6',\n",
       " 'ATP2C2',\n",
       " 'NGFR',\n",
       " 'CDON',\n",
       " 'CTSA',\n",
       " 'SLC12A2',\n",
       " 'EYA2',\n",
       " 'ABCA7',\n",
       " 'SNCAIP',\n",
       " 'FAR2',\n",
       " 'CALCRL',\n",
       " 'ANKS1A',\n",
       " 'TRAM2',\n",
       " 'NTN1',\n",
       " 'MCM10',\n",
       " 'ANKRD44',\n",
       " 'PDIA5',\n",
       " 'GSTO2',\n",
       " 'PRKCQ',\n",
       " 'TLE2',\n",
       " 'FAM107B',\n",
       " 'ME1',\n",
       " 'SLC9A7',\n",
       " 'CTNNA2',\n",
       " 'TIE1',\n",
       " 'ASPM',\n",
       " 'MPPED2',\n",
       " 'FGFR2',\n",
       " 'ZFAT',\n",
       " 'PFKP',\n",
       " 'PLPP1',\n",
       " 'PHKA1',\n",
       " 'PKM',\n",
       " 'PRKCZ',\n",
       " 'SYT1',\n",
       " 'NAV3',\n",
       " 'PDZD4',\n",
       " 'HDAC4',\n",
       " 'REEP1',\n",
       " 'ATP11A',\n",
       " 'PYGM',\n",
       " 'MAST4',\n",
       " 'ADGRF5',\n",
       " 'SDK2',\n",
       " 'ABCC9',\n",
       " 'GAL',\n",
       " 'RORA',\n",
       " 'TGFBR3',\n",
       " 'NEDD4',\n",
       " 'PIGB',\n",
       " 'GUCY2C',\n",
       " 'PTPN3',\n",
       " 'SPTB',\n",
       " 'FGF10',\n",
       " 'SMG6',\n",
       " 'FSTL3',\n",
       " 'ASNS',\n",
       " 'TESK2',\n",
       " 'OSBPL3',\n",
       " 'ATP2B1',\n",
       " 'MGAT4A',\n",
       " 'ARHGAP10',\n",
       " 'RPS6KA2',\n",
       " 'LMCD1',\n",
       " 'MYO3B',\n",
       " 'CYBRD1',\n",
       " 'RPS6KA6',\n",
       " 'LNX1',\n",
       " 'TFRC',\n",
       " 'TRPC5',\n",
       " 'RHOBTB1',\n",
       " 'HMMR',\n",
       " 'TRHDE',\n",
       " 'ACAP1',\n",
       " 'CRMP1',\n",
       " 'IRAG1',\n",
       " 'TMEM38A',\n",
       " 'MCM2',\n",
       " 'TP63',\n",
       " 'PDE8A',\n",
       " 'SDHA',\n",
       " 'ADAM11',\n",
       " 'FERMT2',\n",
       " 'PTGS2',\n",
       " 'ST6GAL1',\n",
       " 'FRY',\n",
       " 'GLI2',\n",
       " 'NOTCH3',\n",
       " 'PPP2R2C',\n",
       " 'CA12',\n",
       " 'MGLL',\n",
       " 'NTN4',\n",
       " 'BCS1L',\n",
       " 'NUAK1',\n",
       " 'SCARF1',\n",
       " 'IPCEF1',\n",
       " 'NOX3',\n",
       " 'TXK',\n",
       " 'WSCD2',\n",
       " 'KCNQ2',\n",
       " 'SEMA3A',\n",
       " 'SEMA3C',\n",
       " 'GRAMD4',\n",
       " 'ADD2',\n",
       " 'FOSL2',\n",
       " 'CACNG5',\n",
       " 'CACNG4',\n",
       " 'ATP12A',\n",
       " 'PAX2',\n",
       " 'EXOSC7',\n",
       " 'MCM6',\n",
       " 'UNG',\n",
       " 'FMO4',\n",
       " 'PLXNA2',\n",
       " 'TPD52',\n",
       " 'PAG1',\n",
       " 'ICAM3',\n",
       " 'MCAM',\n",
       " 'GPC4',\n",
       " 'MBNL3',\n",
       " 'RAP1GAP',\n",
       " 'CTTNBP2',\n",
       " 'RARB',\n",
       " 'PPP1R12B',\n",
       " 'IL4R',\n",
       " 'SNRPA',\n",
       " 'SPAG6',\n",
       " 'APBB1IP',\n",
       " 'GPR137B',\n",
       " 'SLC25A43',\n",
       " 'FBLN1',\n",
       " 'ITGA8',\n",
       " 'MAP2',\n",
       " 'AMPH',\n",
       " 'MCCC1',\n",
       " 'NEBL',\n",
       " 'ADCY2',\n",
       " 'RBFOX1',\n",
       " 'HOXA9',\n",
       " 'EDN1',\n",
       " 'ITM2A',\n",
       " 'TNRC6C',\n",
       " 'CBFA2T2',\n",
       " 'BRINP1',\n",
       " 'RUNX1T1',\n",
       " 'OSBPL6',\n",
       " 'SLC1A3',\n",
       " 'LXN',\n",
       " 'TNS1',\n",
       " 'CDC14A',\n",
       " 'RAPGEF3',\n",
       " 'CEACAM1',\n",
       " 'DUSP13B',\n",
       " 'LIPE',\n",
       " 'OPHN1',\n",
       " 'EPB41L2',\n",
       " 'RIMS1',\n",
       " 'EPHA6',\n",
       " 'RFX3',\n",
       " 'SLC4A4',\n",
       " 'SMARCA2',\n",
       " 'MID2',\n",
       " 'COL5A3',\n",
       " 'KCNN2',\n",
       " 'RBL1',\n",
       " 'CROCCP3',\n",
       " 'NDC80',\n",
       " 'MAGI3',\n",
       " 'CDH7',\n",
       " 'IMPG2',\n",
       " 'MEF2C',\n",
       " 'LRP2',\n",
       " 'CADPS2',\n",
       " 'PCDHA6',\n",
       " 'PHLPP1',\n",
       " 'ATP8B1',\n",
       " 'IL12RB2',\n",
       " 'SMARCD3',\n",
       " 'MPP4',\n",
       " 'C1QTNF3',\n",
       " 'COL19A1',\n",
       " 'EPB41L3',\n",
       " 'DLG3',\n",
       " 'GEMIN5',\n",
       " 'SEMA5B',\n",
       " 'ITGB5',\n",
       " 'TRPM3',\n",
       " 'BCKDHB',\n",
       " 'GRHL2',\n",
       " 'PIBF1',\n",
       " 'TDRD3',\n",
       " 'OXCT1',\n",
       " 'FAT1',\n",
       " 'SLCO1A2',\n",
       " 'NKAIN1',\n",
       " 'EFR3B',\n",
       " 'MECOM',\n",
       " 'MAP3K4',\n",
       " 'ABCB1',\n",
       " 'WNT11',\n",
       " 'RAD54L',\n",
       " 'IPO11',\n",
       " 'NME8',\n",
       " 'EPDR1',\n",
       " 'FAT2',\n",
       " 'ERO1B',\n",
       " 'PPEF1',\n",
       " 'LAT2',\n",
       " 'MYBPC2',\n",
       " 'NOX4',\n",
       " 'ACOX3',\n",
       " 'FTL',\n",
       " 'NLK',\n",
       " 'ADAMTS2',\n",
       " 'TMPRSS11E',\n",
       " 'MMP2',\n",
       " 'GNAO1',\n",
       " 'NID2',\n",
       " 'TFAP2C',\n",
       " 'AURKA',\n",
       " 'PIR',\n",
       " 'TPX2',\n",
       " 'DOCK9',\n",
       " 'DOCK3',\n",
       " 'C3orf18',\n",
       " 'ARHGAP28',\n",
       " 'ATRN',\n",
       " 'SLC4A11',\n",
       " 'LZTS3',\n",
       " 'KIZ',\n",
       " 'DYNLL1',\n",
       " 'TESC',\n",
       " 'SLC23A2',\n",
       " 'DZANK1',\n",
       " 'KDM2B',\n",
       " 'NOS1',\n",
       " 'LAG3',\n",
       " 'ANKRD24',\n",
       " 'LTBP4',\n",
       " 'IRAK3',\n",
       " 'P3H2',\n",
       " 'THPO',\n",
       " 'CHRD',\n",
       " 'CERS4',\n",
       " 'PDPR',\n",
       " 'KIF4A',\n",
       " 'DLL3',\n",
       " 'PITPNM2',\n",
       " 'NRCAM',\n",
       " 'LAMB1',\n",
       " 'SLC26A3',\n",
       " 'ABCC6',\n",
       " 'ITGA6',\n",
       " 'RAPGEF4',\n",
       " 'FH',\n",
       " 'SEL1L3',\n",
       " 'MYO15A',\n",
       " 'PITPNM3',\n",
       " 'ZFHX4',\n",
       " 'ANGPT2',\n",
       " 'CD200',\n",
       " 'CCDC80',\n",
       " 'SLC7A8',\n",
       " 'DAZL',\n",
       " 'SEMA6A',\n",
       " 'TRPM7',\n",
       " 'WDR76',\n",
       " 'TBX15',\n",
       " 'PHGDH',\n",
       " 'COL9A3',\n",
       " 'EZR',\n",
       " 'MYL6',\n",
       " 'CLSPN',\n",
       " 'TGFB2',\n",
       " 'XYLB',\n",
       " 'CDC6',\n",
       " 'PTGS1',\n",
       " 'CRAT',\n",
       " 'SH2D3C',\n",
       " 'WHRN',\n",
       " 'TLL2',\n",
       " 'CYP26A1',\n",
       " 'SORBS1',\n",
       " 'CRTAC1',\n",
       " 'BAMBI',\n",
       " 'IL11',\n",
       " 'MYO3A',\n",
       " 'SMIM24',\n",
       " 'FKBP5',\n",
       " 'EFHC1',\n",
       " 'ITPR3',\n",
       " 'DSP',\n",
       " 'PCSK5',\n",
       " 'ABLIM1',\n",
       " 'NRP1',\n",
       " 'PALMD',\n",
       " 'TSPAN15',\n",
       " 'MAST3',\n",
       " 'MZF1',\n",
       " 'KCNK6',\n",
       " 'CATSPERG',\n",
       " 'HSD3B7',\n",
       " 'PCDH11Y',\n",
       " 'MARCHF2',\n",
       " 'GADD45B',\n",
       " 'PALM',\n",
       " 'MADCAM1',\n",
       " 'ARVCF',\n",
       " 'ZDHHC8',\n",
       " 'DERL3',\n",
       " 'DDTL',\n",
       " 'SF3A1',\n",
       " 'YPEL1',\n",
       " 'LRP5L',\n",
       " 'HIRA',\n",
       " 'SEZ6L',\n",
       " 'LGALS1',\n",
       " 'GCAT',\n",
       " 'TTC28',\n",
       " 'SEPTIN3',\n",
       " 'SLC5A4',\n",
       " 'KDELR3',\n",
       " 'DMC1',\n",
       " 'TIMP3',\n",
       " 'RASL10A',\n",
       " 'AP1B1',\n",
       " 'NEFH',\n",
       " 'MCM5',\n",
       " 'TSPO',\n",
       " 'SYNGR1',\n",
       " 'CACNA1I',\n",
       " 'GRAP2',\n",
       " 'L3MBTL2',\n",
       " 'COCH',\n",
       " 'TRIM9',\n",
       " 'DDHD1',\n",
       " 'DAAM1',\n",
       " 'RIN3',\n",
       " 'LGMN',\n",
       " 'CEP128',\n",
       " 'SLC8A3',\n",
       " 'RPS6KA5',\n",
       " 'C14orf93',\n",
       " 'DHRS2',\n",
       " 'NFKBIA',\n",
       " 'EMC9',\n",
       " 'PLTP',\n",
       " 'MMP9',\n",
       " 'CD40',\n",
       " 'NFATC2',\n",
       " 'SALL4',\n",
       " 'DOK5',\n",
       " 'BMP7',\n",
       " 'PSMA7',\n",
       " 'EEF1A2',\n",
       " 'ISM1',\n",
       " 'RNF24',\n",
       " 'TRIB3',\n",
       " 'RASSF2',\n",
       " 'SNPH',\n",
       " 'CCM2L',\n",
       " 'PLCB4',\n",
       " 'MYL9',\n",
       " 'PAK5',\n",
       " 'MROH8',\n",
       " 'JAG1',\n",
       " 'CDK5RAP1',\n",
       " 'WFDC2',\n",
       " 'PPP1R16B',\n",
       " 'SYNDIG1',\n",
       " 'ACOT8',\n",
       " 'APMAP',\n",
       " 'CELF4',\n",
       " 'ZNF516',\n",
       " 'CDH20',\n",
       " 'ST8SIA5',\n",
       " 'LIPG',\n",
       " 'LAMA1',\n",
       " 'NOL4',\n",
       " 'POLI',\n",
       " 'RBBP8',\n",
       " 'MXRA5',\n",
       " 'ATG4A',\n",
       " 'STS',\n",
       " 'TBL1X',\n",
       " 'POLA1',\n",
       " 'NXT2',\n",
       " 'CHRDL1',\n",
       " 'WDR13',\n",
       " 'MCF2',\n",
       " 'SYP',\n",
       " 'ELF4',\n",
       " 'PGK1',\n",
       " 'PHEX',\n",
       " 'RP2',\n",
       " 'TIMP1',\n",
       " 'KLHL4',\n",
       " 'GABRE',\n",
       " 'PCDH11X',\n",
       " 'KLF8',\n",
       " 'SRPX2',\n",
       " 'RUBCNL',\n",
       " 'NALCN',\n",
       " 'FGF14',\n",
       " 'TNFSF13B',\n",
       " 'FNDC3A',\n",
       " 'ACP5',\n",
       " 'FGF9',\n",
       " 'SGCG',\n",
       " 'FLT1',\n",
       " 'VWA8',\n",
       " 'DGKH',\n",
       " 'KATNAL1',\n",
       " 'INTS6',\n",
       " 'TSNAXIP1',\n",
       " 'PLLP',\n",
       " 'ZNF423',\n",
       " 'DHODH',\n",
       " 'NME3',\n",
       " 'PLA2G15',\n",
       " 'FA2H',\n",
       " 'AXIN1',\n",
       " 'COTL1',\n",
       " 'SLC7A5',\n",
       " 'METRN',\n",
       " 'FBXO31',\n",
       " 'ELOB',\n",
       " 'TOX3',\n",
       " 'QPRT',\n",
       " 'XYLT1',\n",
       " 'PYCARD',\n",
       " 'IL21R',\n",
       " 'SYT17',\n",
       " 'IQCH',\n",
       " 'CORO2B',\n",
       " 'AP3B2',\n",
       " 'ACSBG1',\n",
       " 'CTSH',\n",
       " 'CEMIP',\n",
       " 'ATP8B4',\n",
       " 'OCA2',\n",
       " 'ENTREP2',\n",
       " 'RP1',\n",
       " 'CA2',\n",
       " 'EYA1',\n",
       " 'TRPA1',\n",
       " 'CPQ',\n",
       " 'CALB1',\n",
       " 'SFRP1',\n",
       " 'PLAT',\n",
       " 'JPH1',\n",
       " 'DKK4',\n",
       " 'GDAP1',\n",
       " 'ESRP1',\n",
       " 'STMN2',\n",
       " 'TRPS1',\n",
       " 'NCALD',\n",
       " 'GFUS',\n",
       " 'SH2D4A',\n",
       " 'NEFM',\n",
       " 'ARHGEF10',\n",
       " 'MCM4',\n",
       " 'GYS1',\n",
       " 'MAP4K1',\n",
       " 'RELB',\n",
       " 'PPP1R13L',\n",
       " 'PEX11G',\n",
       " 'AMH',\n",
       " 'RSPH6A',\n",
       " 'IL27RA',\n",
       " 'TNNT1',\n",
       " 'OLFM2',\n",
       " 'RPS16',\n",
       " 'CCNP',\n",
       " 'PLD3',\n",
       " 'SLC1A5',\n",
       " 'APLP1',\n",
       " 'FZR1',\n",
       " 'DMAC2',\n",
       " 'ICAM4',\n",
       " 'ATP1A3',\n",
       " 'MEIS3',\n",
       " 'CNFN',\n",
       " 'PLA2G4C',\n",
       " 'DBP',\n",
       " 'FAM83E',\n",
       " 'RASIP1',\n",
       " 'PLEKHA4',\n",
       " 'PPP2R1A',\n",
       " 'CACNG7',\n",
       " 'PRPF31',\n",
       " 'RPL18A',\n",
       " 'KCNN1',\n",
       " 'ETV2',\n",
       " 'HPN',\n",
       " 'GRIK5',\n",
       " 'CDK6',\n",
       " 'TFPI2',\n",
       " 'PIK3CG',\n",
       " 'ITGB8',\n",
       " 'DNAH11',\n",
       " 'STEAP1B',\n",
       " 'PTN',\n",
       " 'CAV1',\n",
       " 'MET',\n",
       " 'HOXA1',\n",
       " 'HOXA3',\n",
       " 'HOXA5',\n",
       " 'ANKRD7',\n",
       " 'VIPR2',\n",
       " 'TSPAN12',\n",
       " 'HOXA13',\n",
       " 'CPED1',\n",
       " 'CHN2',\n",
       " 'GRB10',\n",
       " 'COBL',\n",
       " 'STX1A',\n",
       " 'NOD1',\n",
       " 'MINDY4',\n",
       " 'HSPB1',\n",
       " 'SNX8',\n",
       " 'PTPRZ1',\n",
       " 'TFR2',\n",
       " 'PCOLCE',\n",
       " 'PPP1R17',\n",
       " 'SERPINE1',\n",
       " 'C1GALT1',\n",
       " 'CLDN15',\n",
       " 'GLCCI1',\n",
       " 'ZNF862',\n",
       " 'MEST',\n",
       " 'POU6F2',\n",
       " 'AHR',\n",
       " 'GLI3',\n",
       " 'AEBP1',\n",
       " 'MYL7',\n",
       " 'CLIP2',\n",
       " 'SLC1A1',\n",
       " 'CNTNAP3',\n",
       " 'PRUNE2',\n",
       " 'MEGF9',\n",
       " 'TRIM14',\n",
       " 'CORO2A',\n",
       " 'TLE4',\n",
       " 'PTGR1',\n",
       " 'SUSD1',\n",
       " 'AKNA',\n",
       " 'DNM1',\n",
       " 'ENG',\n",
       " 'KDM4C',\n",
       " 'DOCK8',\n",
       " 'KANK1',\n",
       " 'ELAVL2',\n",
       " 'RIGI',\n",
       " 'PIP5K1B',\n",
       " 'GLIS3',\n",
       " 'NPDC1',\n",
       " 'APBA1',\n",
       " 'SH3GL2',\n",
       " 'SHB',\n",
       " 'DVL1',\n",
       " 'PDLIM1',\n",
       " 'ATRNL1',\n",
       " 'CXCL12',\n",
       " 'ERLIN1',\n",
       " 'CUBN',\n",
       " 'PALD1',\n",
       " 'UNC5B',\n",
       " 'VSIR',\n",
       " 'SPOCK2',\n",
       " 'BMPR1A',\n",
       " 'ACTA2',\n",
       " 'NPM3',\n",
       " 'ANKRD26',\n",
       " 'MAP3K8',\n",
       " 'DKK1',\n",
       " 'EBF3',\n",
       " 'SORCS1',\n",
       " 'LGI1',\n",
       " 'FBXL20',\n",
       " 'RNF43',\n",
       " 'WNT3',\n",
       " 'TRIM16L',\n",
       " 'HOXB6',\n",
       " 'MED31',\n",
       " 'ICAM2',\n",
       " 'LGALS3BP',\n",
       " 'ASIC2',\n",
       " 'CCL2',\n",
       " 'NAGLU',\n",
       " 'ABI3',\n",
       " 'COL1A1',\n",
       " 'RND2',\n",
       " 'FAM20A',\n",
       " 'MAP2K6',\n",
       " 'MYH3',\n",
       " 'CDR2L',\n",
       " 'PMP22',\n",
       " 'GABRA4',\n",
       " 'NMU',\n",
       " 'CRACD',\n",
       " 'AREG',\n",
       " 'MANBA',\n",
       " 'MAPK10',\n",
       " 'TBC1D9',\n",
       " 'INPP4B',\n",
       " 'GAB1',\n",
       " 'CPE',\n",
       " 'AADAT',\n",
       " 'GALNT7',\n",
       " 'CPZ',\n",
       " 'TRIM2',\n",
       " 'FBXW7',\n",
       " 'SLC75A1',\n",
       " 'GLRB',\n",
       " 'SNX25',\n",
       " 'FAM149A',\n",
       " 'PPARGC1A',\n",
       " 'DDX25',\n",
       " 'CCDC34',\n",
       " 'ZBTB16',\n",
       " 'ELP4',\n",
       " 'HSPA8',\n",
       " 'P2RX3',\n",
       " 'DNAJC4',\n",
       " 'SIAE',\n",
       " 'UNC93B1',\n",
       " 'NRXN2',\n",
       " 'ST3GAL4',\n",
       " 'CCND1',\n",
       " 'CCKBR',\n",
       " 'HPX',\n",
       " 'CEP126',\n",
       " 'GALNT18',\n",
       " 'UPK2',\n",
       " 'KIAA1549L',\n",
       " 'SLC1A2',\n",
       " 'MDK',\n",
       " 'SLC35F2',\n",
       " 'SOX6',\n",
       " 'VWF',\n",
       " 'SELPLG',\n",
       " 'CAPRIN2',\n",
       " 'TSPAN11',\n",
       " 'IL23A',\n",
       " 'SYT10',\n",
       " 'KRT18',\n",
       " 'ACSS3',\n",
       " 'GLI1',\n",
       " 'PPM1H',\n",
       " 'ELK3',\n",
       " 'WNT5B',\n",
       " 'ITFG2',\n",
       " 'PRMT8',\n",
       " 'CUX2',\n",
       " 'AKAP3',\n",
       " 'DUSP16',\n",
       " 'LTBR',\n",
       " 'ARHGDIB',\n",
       " 'VDR',\n",
       " 'ADGRD1',\n",
       " 'TBC1D30',\n",
       " 'ING4',\n",
       " 'CHPT1',\n",
       " 'ENO2',\n",
       " 'PTPN6',\n",
       " 'NT5DC3',\n",
       " 'GOLT1B',\n",
       " 'ST8SIA1',\n",
       " 'CLEC4A',\n",
       " 'COL12A1',\n",
       " 'DSE',\n",
       " 'MAK',\n",
       " 'GCNT2',\n",
       " 'NEDD9',\n",
       " 'MCM9',\n",
       " 'FAM184A',\n",
       " 'MAN1A1',\n",
       " 'TPD52L1',\n",
       " 'RIPOR2',\n",
       " 'SASH1',\n",
       " 'UST',\n",
       " 'ULBP1',\n",
       " 'FBXO5',\n",
       " 'SLC26A8',\n",
       " 'MAPK14',\n",
       " 'SRSF3',\n",
       " 'PHACTR1',\n",
       " 'FBXO9',\n",
       " 'BMP5',\n",
       " 'BACH2',\n",
       " 'GPR63',\n",
       " 'ASCC3',\n",
       " 'RPS12',\n",
       " 'EYA4',\n",
       " 'SOBP',\n",
       " 'SLC16A10',\n",
       " 'ADGRG6',\n",
       " 'PACRG',\n",
       " 'PDE10A',\n",
       " 'SMOC2',\n",
       " 'CCND3',\n",
       " 'BICRAL',\n",
       " 'COX7A2',\n",
       " 'TMEM30A',\n",
       " 'GMDS',\n",
       " 'SLC29A1',\n",
       " 'LAMA4',\n",
       " 'TENT5A',\n",
       " 'CLIC5',\n",
       " 'SEMA5A',\n",
       " 'GHR',\n",
       " 'HMGCS1',\n",
       " 'DAP',\n",
       " 'NME5',\n",
       " 'KIF20A',\n",
       " 'HBEGF',\n",
       " 'SLC4A9',\n",
       " 'LOX',\n",
       " 'SPARC',\n",
       " 'HMGCR',\n",
       " 'HAND1',\n",
       " 'PCDHB7',\n",
       " 'PDE8B',\n",
       " 'RNF130',\n",
       " 'THG1L',\n",
       " 'RASGRF2',\n",
       " 'GABRG2',\n",
       " 'POLR3G',\n",
       " 'CDH6',\n",
       " 'NPR3',\n",
       " 'SLC27A6',\n",
       " 'PDE4D',\n",
       " 'PRLR',\n",
       " 'SLC12A7',\n",
       " 'ST8SIA4',\n",
       " 'PCDH12',\n",
       " 'FGF1',\n",
       " 'NR3C1',\n",
       " 'LIFR',\n",
       " 'C9',\n",
       " 'WWC1',\n",
       " 'DPYSL3',\n",
       " 'PDGFRB',\n",
       " 'CDX1',\n",
       " 'BNIP1',\n",
       " 'STC2',\n",
       " 'CPEB4',\n",
       " 'CNTN3',\n",
       " 'AMOTL2',\n",
       " 'ARMC8',\n",
       " 'RBP1',\n",
       " 'GRK7',\n",
       " 'KAT2B',\n",
       " 'BCHE',\n",
       " 'WNT5A',\n",
       " 'COL7A1',\n",
       " 'FGF12',\n",
       " 'HES1',\n",
       " 'HYAL1',\n",
       " 'NPRL2',\n",
       " 'RPL24',\n",
       " 'GBE1',\n",
       " 'MORC1',\n",
       " 'C3orf52',\n",
       " 'FRMD4B',\n",
       " 'CSPG5',\n",
       " 'CFAP92',\n",
       " 'PLSCR4',\n",
       " 'PEX5L',\n",
       " 'PLCH1',\n",
       " 'ZBTB47',\n",
       " 'FOXP1',\n",
       " 'ZAP70',\n",
       " 'EPB41L5',\n",
       " 'TFCP2L1',\n",
       " 'CYTIP',\n",
       " 'TANC1',\n",
       " 'ITGA4',\n",
       " 'PDE1A',\n",
       " 'PCSK4',\n",
       " 'PCGF1',\n",
       " 'GALNT3',\n",
       " 'TACR1',\n",
       " ...]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata['cNMF'].uns['var_names']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of matching genes: 3844\n",
      "['FAM41C' 'LINC01128' 'LINC00115' ... 'SPRY3' 'VAMP7' 'IL9R']\n",
      "Number of loci: 19853\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Run motif enrichment and save results\n",
    "os.makedirs(output_path, exist_ok=True)\n",
    "for i in range(4):\n",
    "    for class_, thresh in [('enhancer', fimo_thresh_enhancer), \n",
    "                           ('promoter', fimo_thresh_promoter)]:\n",
    "\n",
    "        loci_file = '/oak/stanford/groups/engreitz/Users/ymo/Tools/cNMF_benchmarking/cNMF_benchmarking_pipeline/Evaluation/Resources/scE2G_links/EnhancerPredictionsAllPutative.ForVariantOverlap.shrunk150bp_D{}_{}.tsv'.format(i, class_)\n",
    "        motif_match_df, motif_count_df, motif_enrichment_df = compute_motif_enrichment(\n",
    "            mdata, \n",
    "            prog_key='cNMF',\n",
    "            data_key='rna',\n",
    "            motif_file='/oak/stanford/groups/engreitz/Users/ymo/Tools/cNMF_benchmarking/cNMF_benchmarking_pipeline/Evaluation/Resources/hocomoco_meme.meme',\n",
    "            seq_file='/oak/stanford/groups/engreitz/Users/ymo/Tools/cNMF_benchmarking/cNMF_benchmarking_pipeline/Evaluation/Resources/hg38.fa',\n",
    "            loci_file=loci_file,\n",
    "            window=1000,\n",
    "            sig=thresh,\n",
    "            eps=1e-4,\n",
    "            n_top=2000,\n",
    "            n_jobs=-1,\n",
    "            inplace=False\n",
    "        )\n",
    "\n",
    "        motif_match_df.to_csv(os.path.join(output_path, f'cNMF_{class_}_pearson_topn2000_sample_D{i}_motif_match.txt'), sep='\\t', index=False)\n",
    "        motif_count_df.to_csv(os.path.join(output_path, f'cNMF_{class_}_pearson_topn2000_sample_D{i}_motif_count.txt'), sep='\\t', index=False)\n",
    "        motif_enrichment_df.to_csv(os.path.join(output_path, f'cNMF_{class_}_pearson_topn2000_sample_D{i}_motif_enrichment.txt'), sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "NMF_Benchmarking",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
